auth/dfa.js

```diff
@@ -0,0 +1,605 @@
+ * Will assume roles to be on the form `{roleId, scopes}`.
+ * This will add the property `expandedScopes`, and internal properties
+ * seen, impliedRoles that you should just disregard.
+ *
+ * The resolver returned will take a scope and return a set of scopes granted
+ * by the scope. These will sorted such that they work with `mergeScopeSets`.
+ */
+let computeFixedPoint = (roles) => {
+  // Add initial value for expandedScopes for each role R and sort roles
+  for (let R of roles) {
+    R.expandedScopes = null;
+    R.scopes = sortScopesForMerge(R.scopes)
+    R.impliedRoles = []; // roles that R can directly assume
+    R.seen = 0; // later iteration this role was seen (used later)
+  }
+
+  let {resolver, sets} = buildResolver(roles);
+
+  // Construct impliedRoles
+  for(let R of roles) {
+    let expandImpliedRoles = (index) => {
+      sets[index].forEach(r => {
+        if (typeof(r) === 'number') {
+          expandImpliedRoles(r);
+        } else if (!_.includes(R.impliedRoles, r) && r !== R) {
+          R.impliedRoles.push(r);
+        }
+      });
+    };
+    R.scopes.forEach(scope => expandImpliedRoles(resolver(scope)));
+  }
+
+  // Construct expandedRoles as a fixed-point by traversing implied roles
+  let iteration = 0;
+  let traveseImpliedRoles = (R) => {
+    let scopes = R.scopes;
+    for (let r of R.impliedRoles) {
+      // We traverse all implied roles as long as we haven't already seen them
+      // in this iteration. But just incrementing the `iteration` variable for
+      // each iteration, we don't have to track a list of roles we've seen,
+      // which can be a little expensive as we allocate new memory.
+      if (r.seen < iteration) {
+        r.seen = iteration;
+        if (r.expandedScopes) {
+          // Note that if we've expanded the scopes for a role already, then
+          // there is no reason to traverse this role for scopes as we clearly
+          // have the fixed-point for this role.
+          scopes = mergeScopeSets(scopes, r.expandedScopes);
+        } else {
+          scopes = mergeScopeSets(scopes, traveseImpliedRoles(r));
+        }
+      }
+    }
+    return scopes;
+  };
+  // TODO: make this faster with a simple DFS search and a past-waiting list
+  //       this could probably be significantly faster. Might require some smart
+  //       cycle handling logic, but it seems fairly feasible.
+  //console.time("traveseImpliedRoles");
+  // For each role we compute the fixed-point the set of scopes the role expands
+  // to when all implied roles are considered. If roles are ordered
+  // unfortunately, ie. 1 -> 2 -> 3 -> 4, we can end up traversing the entire
+  // tree N times. Our unit tests is good example of this. To avoid these
+  // degenerate cases we'll quickly shuffle the roles around. In practice such
+  // degenerate cases probably don't exist, but processing order does affect
+  // performance, so we shuffle just for good measure.
```

Comment:
It only handles degenerate cases... we could probably do it smarter... but the whole search could be smarter...
Cost of doing the shuffle is nothing compared to the cost of exploring all the implied roles for each role.